File: common/src/main/java/com/pluxity/device/service/DeviceCategoryService.java
package com.pluxity.device.service;

import static com.pluxity.global.constant.ErrorCode.*;

import com.pluxity.category.service.CategoryService;
import com.pluxity.device.dto.*;
import com.pluxity.device.entity.Device;
import com.pluxity.device.entity.DeviceCategory;
import com.pluxity.device.repository.DeviceCategoryRepository;
import com.pluxity.device.repository.DeviceRepository;
import com.pluxity.facility.Facility;
import com.pluxity.facility.FacilityService;
import com.pluxity.file.dto.FileResponse;
import com.pluxity.file.service.FileService;
import com.pluxity.global.exception.CustomException;
import com.pluxity.global.utils.MappingUtils;
import com.pluxity.global.utils.SortUtils;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import lombok.RequiredArgsConstructor;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class DeviceCategoryService extends CategoryService<DeviceCategory> {

    public static final String DEVICE_CATEGORIES = "device-categories/";
    private final DeviceCategoryRepository deviceCategoryRepository;
    private final DeviceRepository deviceRepository;
    private final FileService fileService;
    private final FacilityService facilityService;

    @Override
    protected JpaRepository<DeviceCategory, Long> getRepository() {
        return deviceCategoryRepository;
    }

    @Transactional
    public Long create(DeviceCategoryRequest request) {
        DeviceCategory deviceCategory =
                DeviceCategory.builder().name(request.name()).iconFileId(request.thumbnailFileId()).build();
        DeviceCategory parent = MappingUtils.findByIdIfExists(request.parentId(), super::findById);

        Long deviceCategoryId = super.create(deviceCategory, parent);

        if (request.thumbnailFileId() != null) {
            deviceCategory.updateIconFileId(request.thumbnailFileId());
            fileService.finalizeUpload(
                    request.thumbnailFileId(), DEVICE_CATEGORIES + deviceCategoryId + "/");
        }

        return deviceCategoryId;
    }

    @Transactional(readOnly = true)
    public DeviceCategoryResponse getDeviceCategory(Long id) {
        List<DeviceCategoryResponse> allCategories = getDeviceCategories();

        return findCategoryInTree(allCategories, id)
                .orElseThrow(() -> new CustomException(NOT_FOUND_DEVICE_CATEGORY, id));
    }

    private Optional<DeviceCategoryResponse> findCategoryInTree(
            List<DeviceCategoryResponse> categories, Long id) {
        for (DeviceCategoryResponse category : categories) {
            if (category.id().equals(id)) {
                return Optional.of(category);
            }

            if (category.children() != null && !category.children().isEmpty()) {
                Optional<DeviceCategoryResponse> foundInChildren =
                        findCategoryInTree(category.children(), id);
                if (foundInChildren.isPresent()) {
                    return foundInChildren;
                }
            }
        }
        return Optional.empty();
    }

    @Transactional(readOnly = true)
    public List<DeviceCategoryResponse> getDeviceCategories() {
        List<DeviceCategory> allCategories =
                deviceCategoryRepository.findAll(SortUtils.getOrderByCreatedAtDesc());

        Map<Long, FileResponse> fileMap =
                MappingUtils.getFileMapByIds(
                        allCategories,
                        deviceCategory -> Stream.of(deviceCategory.getIconFileId()),
                        fileService);

        List<DeviceCategoryResponse> list =
                allCategories.stream()
                        .map(
                                category ->
                                        DeviceCategoryResponse.from(category, fileMap.get(category.getIconFileId())))
                        .toList();

        return MappingUtils.makeCategoryTree(
                list,
                DeviceCategoryResponse::id,
                DeviceCategoryResponse::parentId,
                DeviceCategoryResponse::children);
    }

    @Transactional(readOnly = true)
    public List<DeviceCategoryResponse> getChildDeviceCategories(Long parentId) {
        List<DeviceCategory> childCategories = deviceCategoryRepository.findByParentId(parentId);
        return childCategories.stream()
                .map(this::createDeviceCategoryResponseWithoutChildren)
                .collect(Collectors.toList());
    }

    private DeviceCategoryResponse createDeviceCategoryResponseWithoutChildren(
            DeviceCategory category) {
        FileResponse iconFile =
                category.getIconFileId() != null
                        ? fileService.getFileResponse(category.getIconFileId())
                        : null;
        return DeviceCategoryResponse.fromWithoutChildren(category, iconFile);
    }

    @Transactional
    public void update(Long id, DeviceCategoryUpdateRequest request) {
        DeviceCategory deviceCategory = findById(id);

        if (request.parentId() == null) {
            if (request.name() != null) {
                deviceCategory.updateName(request.name());
            }
            deviceCategory.assignToRootPreservingEntity();
        } else {
            super.update(id, request.name(), request.parentId());
        }

        if (request.thumbnailFileId() != null) {
            deviceCategory.updateIconFileId(request.thumbnailFileId());
            fileService.finalizeUpload(
                    request.thumbnailFileId(), DEVICE_CATEGORIES + deviceCategory.getId() + "/");
        }

        deviceCategory.updateIconFileId(request.thumbnailFileId());
    }

    @Transactional
    public void delete(Long id) {
        DeviceCategory deviceCategory = findById(id);

        if (!deviceCategory.getChildren().isEmpty()) {
            throw new CustomException(CATEGORY_HAS_CHILDREN);
        }

        if (!deviceCategory.getDevices().isEmpty()) {
            throw new CustomException(CATEGORY_HAS_DEVICES);
        }

        deviceCategoryRepository.delete(deviceCategory);
    }

    @Transactional(readOnly = true)
    public List<DeviceCategoryResponse> getRootDeviceCategoryResponses() {
        return getRootCategories().stream()
                .map(this::createDeviceCategoryResponse)
                .collect(Collectors.toList());
    }

    protected DeviceCategoryResponse createDeviceCategoryResponse(DeviceCategory deviceCategory) {
        return DeviceCategoryResponse.from(
                deviceCategory,
                deviceCategory.getIconFileId() != null
                        ? fileService.getFileResponse(deviceCategory.getIconFileId())
                        : null);
    }

    @Transactional(readOnly = true)
    public List<DeviceInfoResponse> getDevicesByCategoryId(Long id, Long facilityId) {
        DeviceCategory category = findById(id);
        Facility facility = facilityService.findById(facilityId);
        List<Device> list = deviceRepository.findByCategoryAndFacility(category, facility);
        return list.stream().map(Device::toDeviceInfo).toList();
    }

    public DeviceCategoryDepthResponse getDeviceCategoryDepth() {
        return new DeviceCategoryDepthResponse(DeviceCategory.builder().build().getMaxDepth());
    }
}

File: common/src/main/kotlin/com/pluxity/device/service/DeviceCategoryService.kt
package com.pluxity.device.service

import com.pluxity.category.dto.CategoryDepthResponse
import com.pluxity.category.service.CategoryService
import com.pluxity.device.dto.DeviceCategoryRequest
import com.pluxity.device.dto.DeviceCategoryResponse
import com.pluxity.device.dto.DeviceCategoryUpdateRequest
import com.pluxity.device.dto.DeviceResponse
import com.pluxity.device.dto.toDeviceCategoryResponse
import com.pluxity.device.dto.toDeviceResponse
import com.pluxity.device.entity.Device
import com.pluxity.device.entity.DeviceCategory
import com.pluxity.device.repository.DeviceCategoryRepository
import com.pluxity.device.repository.DeviceRepository
import com.pluxity.facility.FacilityService
import com.pluxity.file.dto.FileResponse
import com.pluxity.file.extensions.getFileMapById
import com.pluxity.file.service.FileService
import com.pluxity.global.constant.ErrorCode
import com.pluxity.global.exception.CustomException
import com.pluxity.global.utils.MappingUtils
import com.pluxity.global.utils.SortUtils
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class DeviceCategoryService(
    private val deviceCategoryRepository: DeviceCategoryRepository,
    private val deviceRepository: DeviceRepository,
    private val fileService: FileService,
    private val facilityService: FacilityService,
    override val repository: JpaRepository<DeviceCategory, Long>,
) : CategoryService<DeviceCategory>() {
    @Transactional
    fun create(request: DeviceCategoryRequest): Long {
        val deviceCategory =
            DeviceCategory(
                categoryName = request.name,
                iconFileId = request.thumbnailFileId,
            )
        val parent = request.parentId?.let { findById(it) }
        val deviceCategoryId = super.create(deviceCategory, parent)

        request.thumbnailFileId?.let { thumbnailId ->
            deviceCategory.updateIconFileId(thumbnailId)
            fileService.finalizeUpload(
                thumbnailId,
                "$DEVICE_CATEGORIES$deviceCategoryId/",
            )
        }

        return deviceCategoryId
    }

    @Transactional(readOnly = true)
    fun getDeviceCategory(id: Long): DeviceCategoryResponse {
        val allCategories: List<DeviceCategoryResponse> = getDeviceCategories()
        return findCategoryInTree(allCategories, id)
            ?: throw CustomException(ErrorCode.NOT_FOUND_DEVICE_CATEGORY, id)
    }

    private fun findCategoryInTree(
        categories: List<DeviceCategoryResponse>,
        id: Long,
    ): DeviceCategoryResponse? =
        categories.firstOrNull { it.id == id }
            ?: categories
                .asSequence()
                .filter { it.children.isNotEmpty() }
                .mapNotNull { findCategoryInTree(it.children, id) }
                .firstOrNull()

    @Transactional(readOnly = true)
    fun getDeviceCategories(): List<DeviceCategoryResponse> {
        val allCategories: List<DeviceCategory> =
            deviceCategoryRepository.findAllBy(SortUtils.orderByCreatedAtDesc)

        val fileMap = fileService.getFileMapById(allCategories) { it.iconFileId }
        val list: List<DeviceCategoryResponse> =
            allCategories.map { it.toDeviceCategoryResponse(fileMap[it.iconFileId] ?: FileResponse()) }

        return MappingUtils.makeCategoryTree(
            list,
            DeviceCategoryResponse::id,
            DeviceCategoryResponse::parentId,
            DeviceCategoryResponse::children,
        )
    }

    @Transactional(readOnly = true)
    fun getChildDeviceCategories(parentId: Long): List<DeviceCategoryResponse> =
        deviceCategoryRepository.findByParentId(parentId).map { category ->
            createDeviceCategoryResponseWithoutChildren(category)
        }

    private fun createDeviceCategoryResponseWithoutChildren(category: DeviceCategory): DeviceCategoryResponse {
        val iconFile = category.iconFileId?.let { fileService.getFileResponse(it) } ?: FileResponse()
        return category.toDeviceCategoryResponse(iconFile)
    }

    @Transactional
    fun update(
        id: Long,
        request: DeviceCategoryUpdateRequest,
    ) {
        val deviceCategory = findById(id)
        deviceCategory.updateName(request.name)

        // 부모 카테고리 업데이트
        request.parentId?.let { parentId ->
            super.update(id, request.name, parentId)
        } ?: deviceCategory.assignToRootPreservingEntity()

        // 썸네일 파일 업데이트 (중복 제거)
        if (deviceCategory.iconFileId != request.thumbnailFileId) {
            request.thumbnailFileId?.let { thumbnailId ->
                deviceCategory.updateIconFileId(thumbnailId)
                fileService.finalizeUpload(
                    thumbnailId,
                    "$DEVICE_CATEGORIES${deviceCategory.id}/",
                )
            } ?: deviceCategory.updateIconFileId(null)
        }
    }

    @Transactional
    fun delete(id: Long) {
        val deviceCategory = findById(id)

        if (deviceCategory.children.isNotEmpty()) {
            throw CustomException(ErrorCode.CATEGORY_HAS_CHILDREN)
        }

        if (deviceCategory.devices.isNotEmpty()) {
            throw CustomException(ErrorCode.CATEGORY_HAS_DEVICES)
        }

        deviceCategoryRepository.delete(deviceCategory)
    }

    @Transactional(readOnly = true)
    fun getDevicesByCategoryId(
        id: Long,
        facilityId: Long,
    ): List<DeviceResponse> {
        val category = findById(id)
        val facility = facilityService.findById(facilityId)
        val list: List<Device> = deviceRepository.findByCategoryAndFacility(category, facility)
        val fileMap = fileService.getFileMapById(list) { it.category?.iconFileId }
        return list.map { it.toDeviceResponse(fileMap[it.category?.iconFileId]) }
    }

    fun getDeviceCategoryDepth(): CategoryDepthResponse = CategoryDepthResponse(DeviceCategory("").maxDepth)

    companion object {
        const val DEVICE_CATEGORIES: String = "device-categories/"
    }
}
